README
Guake README file
=================

INTRO
~~~~~

Guake is a dropdown terminal made for the GNOME desktop
environment. Its style of window is based on an fps games, and one of
its goals is be easy to reach.

Guake is written mostly in python and has a little piece in C (global
hotkeys stuff). The code is placed in the src directory. Files and
images are in the data directory. Translation stuff is in the po directory.

Features? Bugs? Information?
Visit http://guake.org/


LICENSE
~~~~~~~

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301, USA.


DEPENDENCIES
~~~~~~~~~~~~
 * Python2.4+
 * pygtk2.10 (gtk.StatusIcon)
 * python-vte
 * python-notify
 * python-dbus
 * python-gconf
 * python-xdg

To build guake, you will need the following packages too:
 * python-dev
 * gtk+-2.0-dev
 * pygtk-dev
 * gconf2-dev (to use some autoconf stuff)

Under Debian/Ubuntu, the following command should install all the build
dependencies:
sudo apt-get build-dep guake


INSTALLATION
~~~~~~~~~~~~
We are using an autotools based installation, so if you got the source
of guake from a release tarball, please do the following:

    $ tar -xzvf guake-x.x.x.tar.gz
    $ cd guake-x.x.x
    $ ./configure && make
    $ sudo make install

If you receive a message asking you if you have installed
guake.schemas properly when launching guake, it means that your
default sysconfdir is different from the one chosen by autotools. To
fix that, you probably have to append the param `--sysconfdir=/etc' to
your `./configure' call, like this::

    $ ./configure --sysconfidir=/etc && make

If it is not enought you can call, you can install the gconf schemas
file by hand doing the following::

    # gconftool-2 --makefile-install-rule data/guake.schemas

For more install details, please read the `INSTALL` file.



NEWS
Version 0.8.11
==============

Maintainance release with bug fixes and translation updates.

- #885 revert to the old fixed-width tabs behavior
- move the startup script setting to the hooks tab
- #977 Add a configuration toggle to disable windows refocus
- #970 Right-click tab options don't work properly
- #995 Russian translation
- #983 French translation
- #986 Update German translation


Version 0.8.10
==============

Minors Bug fixes and new Ocean and Oceanic Next color schemes.


Version 0.8.9
=============

Thanks for guakers for the following contibutions:

New features:
- #793, #876: Execute a script on display event
- #864: Add preference dialog checkbox for toggling 'resizer' visibility
- #885: tabs share the full screen width
- #942: Quick open also matches `/home` path
- #933: Add `-l` option to get tab label

Bug Fixes
- #934: Quick open does not work with dash
- #893, #896, #888: another Unity screen size fix
- Translation update: ja (#875), cn (#955), nl (#931), pt (#895),


Version 0.8.8
=============

Thank to these contribution from Guake users, I am happy to announce a new minor fix release of
Guake.

Features:
* Close a tab with the middle button of the mouse

Bug Fixes:
- Fix error when toggle key was disabled
- Update change news
- Uppercase pallete name
- Fix pylint errors
- Convert README badge to SVG
- Update Japanese translation
- update Russian translation
- updated CS translation
- Update zh_CN translation


Version 0.8.7
=============

Do not forget to update the software version

Version 0.8.6
=============

Lot of bug fixes in this release. Thanks for all contributors !

Please note that it is not tested on dual screen set ups.

Bug fixes:
* Terminal geometry fixes (#773 @koter84, #775 RedFlames, b36295 myself)
* Fix "changing max tab length" set all tab to same title
* Fix on terminal kill (#636, @coderstephen)
* Typo/Cosmetics (#761, @thuandt)
* Fix the bottom of tab buttons being cut off in Gnome (#786 @lopsided98)
* Fix fullscreen follow mouse (#774 @koter84)
* Option to shorten VTE tab name (#798 @versusvoid)
* Updated translations:
  - french (b071b4, myself)
  - russian (#787 @vantu5z),
  - corean (#766 @nessunkim),
  - polish (#799 @piotrdrag)



Version 0.8.5
=============

Minor version today, mostly minor bug fixes and translation update.

I did have time to work on GTK3, maintaining Guake to keep using GTK2 is more and more difficult,
Travis kind of abandonned the compatibility of PyGtk2.

* Add a shortcut to open the last tab (#706, thanks @evgenius)
* Fix icon size on Ubuntu (#734)
* Add tab UUID and selection by UUID (#741, thanks @SoniEx2, @Ozzyboshi)
* Updated Polish (#705), Chinese (#711), German (#732), Brazil Portuguese (#744), Czech (#747)
* Fixed doc (#709, #706)
* Fix some Pep8 issue



Version 0.8.4
=============

Bug fixes:
 - Very big icon tray (#598, @thardev)
 - Feature keyboard shorcut hide on lose focus (#650, #262, #350, @thardev)
 - Endless transparency and small rework of hide on lose focus (#651, @thardev)
 - fix tray icon does not align in center (#663, @wuxinyumrx)
 - Updated pt_BR translation (#686, @matheus-manoel)
 - improved Bluloco theme readability (#693, @uloco)
 - ensure gsettings process is well kill (#636)
 - fix exception in preference panel



Version 0.8.3
=============
Quick fix about missing svg file


Version 0.8.2
=============

Bug fix version. Thanks for external contributions!

Feature:
- new palette 'Bluloco' (my default one now!) (@uloco)

Bug fixes:
- tab bar width (@ozzyboshi)
- open new tab in current directory (#578, @Xtreak)
- fix default interpreter (#619, @Xtreak)
- fix use VTE title (#524, @Xtreak)
- Russian tranlation (@vantu5z), german (@Airfunker), spanish (@thardev) chinese (@Xinyu Ng)
- fix guake cannot restore fullscreen (#628, @thardev)


Version 0.8.1
=============

  I started working on Guake 1.0.0, and not in a dedicated branch. It is now in its own source
  folder. We clearly need to move to gtk3 soon, since GTK2 is being discontinued, the VTE is no more
  maintained for GTK2-Python, and adds lot of cool features.

  So I am now starting to work on a complete rewrite of Guake, so don't expect 0.8.x to see lot of
  new features, unfortunately. But Guake 1.0.0 will add features such as:
   - line wrap in terminal
   - search in terminal
   - dconf/gsettings to store configuration
   - GTK3 look and feel
   - much cleaner build and translation systems

  But, this means I cannot work too much on 0.8.x. I still do some bug fixes, and thanks to external
  contributors that share the love for Guake, Guake 0.8 still moves on!

  So don't hesitate to have a look in the code to fix any bug you are experiencing and submit a Pull
  Request.

  New features:
  - a-la guake-indicator custom commands (#564) - by @Ozzyboshi!
  - Add option to allow/disallow bold font (#603) - by @helix84!
  - Clean current terminal item in contextual menu (#608) - by @Denis Subbotin

  Bug fixes:
  - Terminal widget disappears at random times (#592)
  - Typo - by @selivan, @Ruined1


Version 0.8.0
=============

  I have been extremely busy the previous 3 months, so I have almost not worked on Guake. I wanted
  to introduce in the next version some major features heavily asked, like session save and split
  terminal. They will have to wait a bit more.

  As a result, most of the contribution are from external contributors. Thank you very much for all
  these patches!

  This releases introduces two major changes in the project, thus the minor version change.

  First, the new homepage is now online:

    http://guake-project.org/

  As I remind you, Guake has *not* control over the old domain guake.org. So far the content is
  still one of the old content of this domain. So please use http://guake-project.org to reference
  Guake.

  Source code of the Web site can be found here:

    https://github.com/Guake/guake-website

  The second major change in the project is the abandon of our internal hotkey manager

guake/main.py
# -*- coding: utf-8; -*-
"""
Copyright (C) 2007-2013 Guake authors

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public
License along with this program; if not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301 USA
"""

import gi
gi.require_version('Gtk', '3.0')

from gi.repository import Gtk

import dbus
import logging
import os
import subprocess
import sys
import uuid

from optparse import OptionParser

from guake.common import _

from guake.globals import VERSION

from guake.dbusiface import DBUS_NAME
from guake.dbusiface import DBUS_PATH
from guake.dbusiface import DbusManager

from guake.guake_app import Guake
from guake.guake_logging import setupLogging

log = logging.getLogger(__name__)


def main():
    """Parses the command line parameters and decide if dbus methods
    should be called or not. If there is already a guake instance
    running it will be used and a True value will be returned,
    otherwise, false will be returned.
    """

    setupLogging(True)

    # COLORTERM is an environment variable set by some terminal emulators such as gnome-terminal.
    # To avoid confusing applications running inside Guake, clean up COLORTERM at startup.
    if "COLORTERM" in os.environ:
        del os.environ['COLORTERM']

    # Force to xterm-256 colors for compatibility with some old command line programs
    os.environ["TERM"] = "xterm-256color"

    parser = OptionParser(version='Guake Terminal %s' % VERSION)
    parser.add_option(
        '-f',
        '--fullscreen',
        dest='fullscreen',
        action='store_true',
        default=False,
        help=_('Put Guake in fullscreen mode')
    )

    parser.add_option(
        '-t',
        '--toggle-visibility',
        dest='show_hide',
        action='store_true',
        default=False,
        help=_('Toggles the visibility of the terminal window')
    )

    parser.add_option(
        '--show',
        dest="show",
        action='store_true',
        default=False,
        help=_('Shows Guake main window')
    )

    parser.add_option(
        '--hide',
        dest='hide',
        action='store_true',
        default=False,
        help=_('Hides Guake main window')
    )

    parser.add_option(
        '-p',
        '--preferences',
        dest='show_preferences',
        action='store_true',
        default=False,
        help=_('Shows Guake preference window')
    )

    parser.add_option(
        '-a',
        '--about',
        dest='show_about',
        action='store_true',
        default=False,
        help=_('Shows Guake\'s about info')
    )

    parser.add_option(
        '-n',
        '--new-tab',
        dest='new_tab',
        action='store',
        default='',
        help=_('Add a new tab (with current directory set to NEW_TAB)')
    )

    parser.add_option(
        '-s',
        '--select-tab',
        dest='select_tab',
        action='store',
        default='',
        help=_('Select a tab (SELECT_TAB is the index of the tab)')
    )

    parser.add_option(
        '-g',
        '--selected-tab',
        dest='selected_tab',
        action='store_true',
        default=False,
        help=_('Return the selected tab index.')
    )

    parser.add_option(
        '-l',
        '--selected-tablabel',
        dest='selected_tablabel',
        action='store_true',
        default=False,
        help=_('Return the selected tab label.')
    )

    parser.add_option(
        '-e',
        '--execute-command',
        dest='command',
        action='store',
        default='',
        help=_('Execute an arbitrary command in the selected tab.')
    )

    parser.add_option(
        '-i',
        '--tab-index',
        dest='tab_index',
        action='store',
        default='0',
        help=_('Specify the tab to rename. Default is 0.')
    )

    parser.add_option(
        '--bgcolor',
        dest='bgcolor',
        action='store',
        default='',
        help=_('Set the hexadecimal (#rrggbb) background color of '
               'the selected tab.')
    )

    parser.add_option(
        '--fgcolor',
        dest='fgcolor',
        action='store',
        default='',
        help=_('Set the hexadecimal (#rrggbb) foreground color of the '
               'selected tab.')
    )

    parser.add_option(
        '--rename-tab',
        dest='rename_tab',
        metavar='TITLE',
        action='store',
        default='',
        help=_('Rename the specified tab. Reset to default if TITLE is '
               'a single dash "-".')
    )

    parser.add_option(
        '-r',
        '--rename-current-tab',
        dest='rename_current_tab',
        metavar='TITLE',
        action='store',
        default='',
        help=_('Rename the current tab. Reset to default if TITLE is a '
               'single dash "-".')
    )

    parser.add_option(
        '-q',
        '--quit',
        dest='quit',
        action='store_true',
        default=False,
        help=_('Says to Guake go away =(')
    )

    parser.add_option(
        '-u',
        '--no-startup-script',
        dest='execute_startup_script',
        action='store_false',
        default=True,
        help=_('Do not execute the start up script')
    )

    options = parser.parse_args()[0]

    instance = None

    # Trying to get an already running instance of guake. If it is not
    # possible, lets create a new instance. This function will return
    # a boolean value depending on this decision.
    try:
        bus = dbus.SessionBus()
        remote_object = bus.get_object(DBUS_NAME, DBUS_PATH)
        already_running = True
    except dbus.DBusException:
        log.debug("DBus not running, starting it")
        instance = Guake()
        remote_object = DbusManager(instance)
        already_running = False

    only_show_hide = True

    if options.fullscreen:
        remote_object.fullscreen()

    if options.show:
        remote_object.show_from_remote()

    if options.hide:
        remote_object.hide_from_remote()

    if options.show_preferences:
        remote_object.show_prefs()
        only_show_hide = False

    if options.new_tab:
        remote_object.add_tab(options.new_tab)
        only_show_hide = False

    if options.select_tab:
        selected = int(options.select_tab)
        i = remote_object.select_tab(selected)
        if i is None:
            sys.stdout.write('invalid index: %d\n' % selected)
        only_show_hide = False

    if options.selected_tab:
        selected = remote_object.get_selected_tab()
        sys.stdout.write('%d\n' % selected)
        only_show_hide = False

    if options.selected_tablabel:
        selectedlabel = remote_object.get_selected_tablabel()
        sys.stdout.write('%s\n' % selectedlabel)
        only_show_hide = False

    if options.command:
        remote_object.execute_command(options.command)
        only_show_hide = False

    if options.tab_index and options.rename_tab:
        try:
            remote_object.rename_tab_uuid(str(uuid.UUID(options.tab_index)), options.rename_tab)
        except ValueError:
            remote_object.rename_tab(int(options.tab_index), options.rename_tab)
        only_show_hide = False

    if options.bgcolor:
        remote_object.set_bgcolor(options.bgcolor)
        only_show_hide = False

    if options.fgcolor:
        remote_object.set_fgcolor(options.fgcolor)
        only_show_hide = False

    if options.rename_current_tab:
        remote_object.rename_current_tab(options.rename_current_tab)
        only_show_hide = False

    if options.show_about:
        remote_object.show_about()
        only_show_hide = False

    if options.quit:
        try:
            remote_object.quit()
            return True
        except dbus.DBusException:
            return True

    if already_running and only_show_hide:
        # here we know that guake was called without any parameter and
        # it is already running, so, lets toggle its visibility.
        remote_object.show_hide()

    # TODO PORT remove the next line it only exists for testing...
    remote_object.show_hide()
    # TODO PORT remove the next line it only exists for testing...
    # remote_object.show_prefs()

    if options.execute_startup_script:
        if not already_running:
            startup_script = instance.settings.general.get_string("startup-script")
            if startup_script:
                log.info("Calling startup script: %s", startup_script)
                pid = subprocess.Popen([startup_script],
                                       shell=True,
                                       stdin=None,
                                       stdout=None,
                                       stderr=None,
                                       close_fds=True)
                log.info("Startup script started with pid: %s", pid)
                # Please ensure this is the last line !!!!
    else:
        log.info("--no-startup-script argument defined, so don't execute the startup script")
    if already_running:
        log.info("Guake is already running")
    return already_running


def exec_main():
    if not main():
        log.info("Running main gtk loop")
        Gtk.main()


if __name__ == '__main__':
    exec_main()
